tower-http/src/services/fs/mod.rs

```diff
@@ -213,3 +214,76 @@ fn response_from_io_error(
         _ => Err(err),
     }
 }
+
+struct LastModified(HttpDate);
+
+impl From<SystemTime> for LastModified {
+    fn from(time: SystemTime) -> Self {
+        LastModified(time.into())
+    }
+}
+
+struct IfUnmodifiedSince(HttpDate);
+struct IfModifiedSince(HttpDate);
+
+impl IfModifiedSince {
+    /// Check if the supplied time means the resource has been modified.
+    fn is_modified(&self, last_modified: &LastModified) -> bool {
+        self.0 < last_modified.0
+    }
+
+    /// convert a header value into a IfModifiedSince, invalid values are silentely ignored
+    fn from_header_value(value: &http::header::HeaderValue) -> Option<IfModifiedSince> {
+        String::from_utf8(value.as_bytes().to_vec())
+            .ok()
+            .map(|value| httpdate::parse_http_date(&value).ok())
+            .flatten()
+            .map(|time| IfModifiedSince(time.into()))
+    }
+}
+
+impl IfUnmodifiedSince {
+    /// Check if the supplied time passes the precondtion.
+    fn precondition_passes(&self, last_modified: &LastModified) -> bool {
+        self.0 >= last_modified.0
+    }
+
+    /// convert a header value into a IfModifiedSince, invalid values are silentely ignored
+    fn from_header_value(value: &http::header::HeaderValue) -> Option<IfUnmodifiedSince> {
+        String::from_utf8(value.as_bytes().to_vec())
+            .ok()
+            .map(|value| httpdate::parse_http_date(&value).ok())
+            .flatten()
+            .map(|time| IfUnmodifiedSince(time.into()))
+    }
+}
+
+fn check_modified_headers(
+    modified: &Option<LastModified>,
```

Comment:
Usually you don’t want references to options, but an optional reference. `Option::as_ref` Can be used

Reply:
in that case this does not matter, but I agree.